agent/grpc-external/services/dns/server.go

```diff
@@ -0,0 +1,136 @@
+package dns
+
+import (
+	"context"
+	"fmt"
+	"net"
+
+	"github.com/hashicorp/go-hclog"
+	"github.com/miekg/dns"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/peer"
+	"google.golang.org/grpc/status"
+
+	"github.com/hashicorp/consul/proto-public/pbdns"
+)
+
+type Local struct {
+	IP   net.IP
+	Port int
+}
+
+type Config struct {
+	Logger       hclog.Logger
+	DNSServeMux  *dns.ServeMux
+	LocalAddress Local
+}
+
+type Server struct {
+	Config
+}
+
+func NewServer(cfg Config) *Server {
+	return &Server{cfg}
+}
+
+func (s *Server) Register(grpcServer *grpc.Server) {
+	pbdns.RegisterDNSServiceServer(grpcServer, s)
+}
+
+// BufferResponseWriter writes a DNS response to a byte buffer.
+type BufferResponseWriter struct {
+	ResponseBuffer []byte
+	LocalAddress   net.Addr
+	RemoteAddress  net.Addr
+	Logger         hclog.Logger
+}
+
+// LocalAddr returns the net.Addr of the server
+func (b *BufferResponseWriter) LocalAddr() net.Addr {
+	return b.LocalAddress
+}
+
+// RemoteAddr returns the net.Addr of the client that sent the current request.
+func (b *BufferResponseWriter) RemoteAddr() net.Addr {
+	return b.RemoteAddress
+}
+
+// WriteMsg writes a reply back to the client.
+func (b *BufferResponseWriter) WriteMsg(m *dns.Msg) error {
+	// Pack message to bytes first.
+	msgBytes, err := m.Pack()
+	if err != nil {
+		b.Logger.Error("error packing message", "err", err)
+		return err
+	}
+	b.ResponseBuffer = msgBytes
+	return nil
+}
+
+// Write writes a raw buffer back to the client.
+func (b *BufferResponseWriter) Write(m []byte) (int, error) {
+	b.Logger.Info("Write was called")
+	copy(b.ResponseBuffer, m)
+	return len(b.ResponseBuffer), nil
```

Comment:
`copy` returns the number of elements copied, so you don't have to count the buffer twice:
```suggestion
	return copy(b.ResponseBuffer, m), nil
```